common/client/transaction_sender.go

```diff
@@ -0,0 +1,261 @@
+// The aggregation is based on the following conditions:
+// * If there is at least one success - returns success
+// * If there is at least one terminal error - returns terminal error
+// * If there is both success and terminal error - returns success and reports invariant violation
+// * Otherwise, returns any (effectively random) of the errors.
+func (txSender *transactionSender[TX, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) (SendTxReturnCode, error) {
+	txResults := make(chan sendTxResult, len(txSender.multiNode.primaryNodes))
+	txResultsToReport := make(chan sendTxResult, len(txSender.multiNode.primaryNodes))
+	primaryWg := sync.WaitGroup{}
+
+	var err error
+	ok := txSender.multiNode.IfNotStopped(func() {
+		err = txSender.multiNode.DoAll(ctx, func(ctx context.Context, rpc RPC, isSendOnly bool) bool {
+			if isSendOnly {
+				// Use multiNode wg to ensure transactions are done sending before multinode shuts down
+				txSender.multiNode.wg.Add(1)
+				fmt.Println("Calling send only rpc SendTransaction()")
+				go func() {
+					defer txSender.multiNode.wg.Done()
+					// Send-only nodes' results are ignored as they tend to return false-positive responses.
+					// Broadcast to them is necessary to speed up the propagation of TX in the network.
+					_ = txSender.broadcastTxAsync(ctx, rpc, tx)
+				}()
+				return true
+			}
+
+			// Primary Nodes
+			primaryWg.Add(1)
+			go func() {
+				defer primaryWg.Done()
+				result := txSender.broadcastTxAsync(ctx, rpc, tx)
+				txResultsToReport <- result
+				txResults <- result
+			}()
+			return true
+		})
+		if err != nil {
+			primaryWg.Wait()
+			close(txResultsToReport)
+			close(txResults)
+			return
+		}
+
+		// This needs to be done in parallel so the reporting knows when it's done (when the channel is closed)
+		txSender.multiNode.wg.Add(1)
+		go func() {
+			defer txSender.multiNode.wg.Done()
+			primaryWg.Wait()
+			close(txResultsToReport)
+			close(txResults)
+		}()
+
+		txSender.multiNode.wg.Add(1)
+		go txSender.reportSendTxAnomalies(tx, txResultsToReport)
+	})
+	if !ok {
+		return 0, fmt.Errorf("aborted while broadcasting tx - MultiNode is stopped: %w", context.Canceled)
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	return txSender.collectTxResults(ctx, tx, len(txSender.multiNode.primaryNodes), txResults)
+}
+
+func (txSender *transactionSender[TX, CHAIN_ID, RPC]) broadcastTxAsync(ctx context.Context, rpc RPC, tx TX) sendTxResult {
+	txErr := rpc.SendTransaction(ctx, tx)
+	txSender.lggr.Debugw("Node sent transaction", "tx", tx, "err", txErr)
+	resultCode := txSender.txErrorClassifier(tx, txErr)
+	if !slices.Contains(sendTxSuccessfulCodes, resultCode) {
+		txSender.lggr.Warnw("RPC returned error", "tx", tx, "err", txErr)
+	}
+	return sendTxResult{Err: txErr, ResultCode: resultCode}
+}
+
+func (txSender *transactionSender[TX, CHAIN_ID, RPC]) reportSendTxAnomalies(tx TX, txResults <-chan sendTxResult) {
+	defer txSender.multiNode.wg.Done()
```

Comment:
Yeah I was wondering about this. The previous implementation was optimized to use the multinode waitgroup for sendOnly transactions since they need to finish before multiNode shuts down, but not necessarily during the executions of these functions. However, I think we should probably just synchronize inside of these functions instead for simplicity and reliability.